lesson20.md

```diff
@@ -0,0 +1,319 @@
+<!-- .slide: id="lesson20" -->
+
+# Basic Frontend - Spring 2021
+
+Lesson 20, Tuesday, 2021-05-18
+
+---
+
+### Recap - Data types
+
+```js
+let a1 = 42;
+let a2 = true;
+let a3 = "hello";
+let a4 = function() {};
+let a5;
+let a6 = [];
+let a7 = {};
+```
+
+---
+
+### Solution - Data types
+
+```js
+let a1 = 42;       // number
+let a2 = true;     // boolean
+let a3 = "hello";  // string
+let a4 = function() {};  // function
+let a5;            // undefined
+let a6 = [];       // object (array)
+let a7 = {};       // object
+```
+
+---
+
+### Null
+
+* There's one datatype that we haven't learned about yet: `null`
+* `null` is like `undefined`. It indicates the absence of a value.
+* `null` can be used to indicate an *intentional* absence of an object value.
+* Example: `document.getElementById("NoExist");` returns `null`
+
+---
+
+### Recap - DOM
+
+* DOM is an API
+* It allows us to add/remove/change any element of a web page
+* We can make TODO lists and cool web shops
+* But how would we create a web page with "dynamic" data, for example today's weather?
+
+---
+
+<!-- .slide: id="WebAPI" -->
+
+# Web APIs
+
+---
+
+### Basic building blocks
+
+We need to first learn about a few concepts:
+
+1. JSON - How to exchange data
+1. REST API - How to talk with a server
+1. `fetch` API - How to connect to a server
+
+---
+
+### Object notation
+
+In JavaScript, we can put all keys of objects in quotes. This is purely optional:
+
+```js
+let me = {
+    "name": "John",
+    lastName: "Doe",
+    "hobbies": [ "Eat", "Sleep" ]
+};
+
+console.log(me.name); // "John"
+```
+
+---
+
+### JSON
+
+* JSON stands for **J**ava**S**cript **O**bject **N**otation
+* It is a string representation of a JavaScript object
+* All *keys* in our key value pairs must be in quotes
+* Must contain only these types:
+   * `string`, `number`, `boolean`, `array`
+   * `null`
+   * another JSON `object`
+* These types are **forbidden**:
+   * `function`, `undefined`
+
+---
+
+### Why JSON?
+
+* When we exchange data between a browser and a server, we can only exchange data, for example a string (not objects, arrays, numbers, booleans...)
+* JSON is a string representation of JavaScript objects
+* JSON can be easily transferred (and stored)
+* The notation is very close to JavaScript objects, easy to handle from JavaScript. No complex transformation required.
+* More info: https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/JSON
+
+---
+
+### JSON example
+
+```json
+{
+    "name": "Bob",
+    "instructor": true,
+    "hobbies": [ "eat", "sleep" ],
+    "age": 86,
+    "workAddress": {
+        "street": "Invalidenstraße",
+        "number": 116,
+        "city": "Berlin"
+    }
+}
+```
+
+---
+
+### object <-> JSON
```

Comment:
Skip JSON.stringify and JSON.parse API, we don't make use of them in this lesson at all. Last semester it got students confused.